fem/nonlininteg.cpp

```diff
@@ -385,4 +437,226 @@ void HyperelasticNLFIntegrator::AssembleElementGrad(const FiniteElement &el,
+   *elmats(0,1) = 0.0;
+   *elmats(1,0) = 0.0;
+   *elmats(1,1) = 0.0;
+
+   DSh_u.SetSize(dof_u, dim);
+   DS_u.SetSize(dof_u, dim);
+   J0i.SetSize(dim);
+   F.SetSize(dim);
+   FinvT.SetSize(dim);
+   Finv.SetSize(dim);
+   P.SetSize(dim);
+   PMatI_u.UseExternalData(elfun[0]->GetData(), dof_u, dim);
+   Sh_p.SetSize(dof_p);
+
+   int intorder = 2*el[0]->GetOrder() + 3; // <---
+   const IntegrationRule &ir = IntRules.Get(el[0]->GetGeomType(), intorder);
+
+   for (int i = 0; i < ir.GetNPoints(); i++)
+   {
+      const IntegrationPoint &ip = ir.IntPoint(i);
+      Tr.SetIntPoint(&ip);
+      CalcInverse(Tr.Jacobian(), J0i);
+
+      el[0]->CalcDShape(ip, DSh_u);
+      Mult(DSh_u, J0i, DS_u);
+      MultAtB(PMatI_u, DS_u, F);
+
+      el[1]->CalcShape(ip, Sh_p);
+      double pres = Sh_p * *elfun[1];
+      double mu = c_mu->Eval(Tr, ip);
+      double dJ = F.Det();
+
+      CalcInverseTranspose(F, FinvT);
+
+      // u,u block
+      for (int i_u = 0; i_u < dof_u; i_u++)
+      {
+         for (int i_dim = 0; i_dim < dim; i_dim++)
+         {
+            for (int j_u = 0; j_u < dof_u; j_u++)
+            {
+               for (int j_dim = 0; j_dim < dim; j_dim++)
+               {
+
+                  // m = j_dim;
+                  // k = i_dim;
+
+                  for (int n=0; n<dim; n++)
+                  {
+                     for (int l=0; l<dim; l++)
+                     {
+
+                        (*elmats(0,0))(i_u + i_dim*dof_u, j_u + j_dim*dof_u) += dJ * (mu * F(i_dim,
+                                                                                             l) - pres * FinvT(i_dim,l)) * FinvT(j_dim,n) * DS_u(i_u,l) * DS_u(j_u,
+                                                                                                   n) * ip.weight * Tr.Weight();
+
+                        if (j_dim == i_dim && n==l)
+                        {
+                           (*elmats(0,0))(i_u + i_dim*dof_u, j_u + j_dim*dof_u) += dJ * mu * DS_u(i_u,
+                                                                                                  l) * DS_u(j_u,n) * ip.weight * Tr.Weight();
+                        }
+
+                        // a = n;
+                        // b = m;
+                        (*elmats(0,0))(i_u + i_dim*dof_u, j_u + j_dim*dof_u) += dJ * pres * FinvT(i_dim,
+                                                                                                  n) * FinvT(j_dim,l) * DS_u(i_u,l) * DS_u(j_u,n) * ip.weight * Tr.Weight();
+                     }
+                  }
+               }
+            }
+         }
+      }
+
+      // u,p and p,u blocks
+      for (int i_p = 0; i_p < dof_p; i_p++)
+      {
+         for (int j_u = 0; j_u < dof_u; j_u++)
+         {
+            for (int dim_u = 0; dim_u < dim; dim_u++)
+            {
+               for (int l=0; l<dim; l++)
+               {
+                  (*elmats(1,0))(i_p, j_u + dof_u * dim_u) += dJ * FinvT(dim_u,l) * DS_u(j_u,
```

Comment:
You can save a little computation here by computing dJ*FinvT*DS_u*W once and using it for both elmats(0,1) and elmats(1,0).

Reply:
Ah yes! I remember I was going to do some optimizations in these compute-heavy loops and I totally forgot!
- [ ] Optimize gradient calculation